projects/user_publication/gscholar.py

```diff
@@ -0,0 +1,207 @@
+        self.scholarly.use_proxy(self.pg)
+        self.retries = 0
+        self.citations = []
+
+    def _handle_proxy_reload(func):
+        def inner_f(self, *args, **kwargs):
+            while self.retries < MAX_RETRIES:
+                try:
+                    logger.info(f"Using Free proxy - {True if self.pg.has_proxy() else False}")
+                    resp = func(self, *args, **kwargs)
+                except MaxTriesExceededException:
+                    # this occurs if Google blocks IP
+                    logger.info(f"New free proxy retries : {self.retries} / {MAX_RETRIES}")
+                    self.retries += 1
+                    self.new_proxy()
+                else:
+                    self.retries = 0
+                    break
+            return resp
+        return inner_f
+
+    def _publication_id(self, pub: dict):
+        m = re.search(r"cites=[\d+,]*", pub["citedby_url"])
+        return m.group()[6:]
+
+    def new_proxy(self):
+        """sets a new proxy from free-proxy library"""
+        if not self.pg.proxy_mode.value == "FREE_PROXIES":
+            return
+        logger.info("Getting new free proxy")
+        self.pg.get_next_proxy()
+        self.scholarly.use_proxy(self.pg)
+
+    @_handle_proxy_reload
+    def get_one_pub(self, title: str):
+        """To get one publication from google scholar
+        Make sure the title is an exact match
+
+        Args:
+            title (str): Title of the publication
+        """
+        return self.scholarly.search_single_pub(title)
+
+    @_handle_proxy_reload
+    def get_cites(
+            self, pub, year_low=2014,
+            year_high=datetime.now().year
+    ):
+        """Returns all the citations of the publication
+
+        Args:
+            pub (dict): return of the search_single_pub method
+            year_low (int, optional): year to query from. Defaults to 2014
+            year_high (int, optional): yaar to query till. Defaults to datetime.now().year
+
+        Returns:
+            list: list of publications that cited arg:pub
+        """
+        pub_id = self._publication_id(pub)
+        logger.info(f"Getting citations for {pub['bib']['title']}")
+        no_citations = pub['num_citations']
+        cites_gen = self.scholarly.search_citedby(
+            pub_id, year_low=year_low, year_high=year_high
+        )
+        citations = []
+        cit_count = 0
+        while True:
+            try:
+                cited_pub = self._get_bib(self._get_next_cite(cites_gen))
+                citations.append(cited_pub)
+                cit_count += 1
+                logger.info(f"Got {len(cit_count)} / {no_citations}")
+            except StopIteration:
+                logger.info(f"End of iteration. Got {len(citations)} / {no_citations}")
+                return citations
+
+    @_handle_proxy_reload
+    def _get_next_cite(self, cites_gen):
+        return next(cites_gen)
+
+    @_handle_proxy_reload
+    def _get_bib(self, pub):

```

Comment:
This sounds like it should fetch the bibtex from the pub and return the bibtex, but it actually fetches the bibtex, adds it to the pub, and returns the pub.